File: giapi-status-service/src/main/java/edu/gemini/aspen/giapi/statusservice/AbstractStatusItemTranslator.java
package edu.gemini.aspen.giapi.statusservice;

import com.google.common.base.Optional;
import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;
import edu.gemini.aspen.giapi.status.Health;
import edu.gemini.aspen.giapi.status.StatusItem;
import edu.gemini.aspen.giapi.status.impl.BasicStatus;
import edu.gemini.aspen.giapi.status.impl.HealthStatus;
import edu.gemini.aspen.giapi.statusservice.generated.DataType;
import edu.gemini.aspen.giapi.statusservice.generated.MapType;
import edu.gemini.aspen.giapi.statusservice.generated.StatusType;
import edu.gemini.aspen.giapi.util.jms.status.StatusGetter;
import edu.gemini.gmp.top.Top;
import net.jmatrix.eproperties.EProperties;
import org.xml.sax.SAXException;

import javax.xml.bind.JAXBException;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.*;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Base class for building status translators
 */
abstract public class AbstractStatusItemTranslator implements StatusItemTranslator {
    private static final Logger LOG = Logger.getLogger(AbstractStatusItemTranslator.class.getName());
    private static final String CONF_DIR_PROPERTY = "statusTranslatorFile";


    /**
     * This class is needed to allow for translations of one StatusItem mapping to multiple ones,
     * as well as several different items mapping to a single one.
     */
    private static class Key {
        private String original;
        private String translated;

        static Key create(String original, String translated) {
            return new Key(original, translated);
        }

        private Key(String original, String translated) {
            this.original = original;
            this.translated = translated;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Key key = (Key) o;

            if (!original.equals(key.original)) return false;
            if (!translated.equals(key.translated)) return false;

            return true;
        }

        @Override
        public int hashCode() {
            int result = original.hashCode();
            result = 31 * result + translated.hashCode();
            return result;
        }
    }

    /**
     * Maps original StatusItem name to 1 or more translated names
     */
    private final Multimap<String, String> names = HashMultimap.create();
    /**
     * Maps the name of the translation to its type
     */
    private final Map<Key, DataType> types = new HashMap<Key, DataType>();
    /**
     * Maps the name of a translation to the actual translations
     */
    private final Map<Key, Map<String, String>> translations = new HashMap<Key, Map<String, String>>();
    /**
     * Maps the name of a translation to its default values
     */
    private final Map<Key, String> defaults = new HashMap<Key, String>();

    private final String xmlFileName;
    private final String name = "StatusItemTranslator: " + this;
    protected final Top top;
    protected StatusItemTranslatorConfiguration config;
    protected final StatusGetter getter = new StatusGetter("Status Translator initial item loader");
    protected final AtomicBoolean jmsStarted = new AtomicBoolean(false);

    public AbstractStatusItemTranslator(Top top, String xmlFileName) {
        this.top = top;
        this.xmlFileName = xmlFileName;
    }

    static protected void waitFor(AtomicBoolean bool) {
        long sleepTime = 100;
        do {
            if (!bool.get()) {
                try {
                    Thread.sleep(sleepTime *= 2);
                } catch (InterruptedException e) {
                    LOG.log(Level.SEVERE, e.getMessage(), e);
                }
            }
        } while (!bool.get());
    }

    /**
     * Reads in the configuration and stores it for using it in  translations
     *
     * @throws IOException
     * @throws JAXBException
     */
    public void start() throws IOException, JAXBException, SAXException {
        File f = new File(substituteProperties(xmlFileName));
        if (!f.exists()) {
            LOG.severe("Configuration file " + f + " does not exist");
            return;
        }
        LOG.info("Start StatusItemTranslator with configuration file " + f);

        //read mappings
        config = new StatusItemTranslatorConfiguration(new FileInputStream(f));

        // initialize mappings
        for (StatusType status : config.getStatuses()) {
            Map<String, String> tr = new HashMap<String, String>();
            for (MapType map : status.getMaps().getMap()) {
                tr.put(map.getFrom(), map.getTo());
            }
            translations.put(Key.create(top.buildStatusItemName(status.getOriginalName()), top.buildStatusItemName(status.getTranslatedName())), tr);
            if (status.getDefault() != null) {
                defaults.put(Key.create(top.buildStatusItemName(status.getOriginalName()), top.buildStatusItemName(status.getTranslatedName())), status.getDefault());
            }
        }

        //store types and names
        for (StatusType status : config.getStatuses()) {
            types.put(Key.create(top.buildStatusItemName(status.getOriginalName()), top.buildStatusItemName(status.getTranslatedName())), status.getTranslatedType());
            names.put(top.buildStatusItemName(status.getOriginalName()), top.buildStatusItemName(status.getTranslatedName()));
        }
    }

    /**
     * Try to fetch items from the StatusDB at startup. Translate those found.
     */
    protected void initItems() {
        ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(1);
        executor.execute(new Runnable() {
            @Override
            public void run() {
                LOG.fine("Start initItems");
                try {
                    waitFor(jmsStarted);//wait until JMS is started

                    long sleepTime = 100;
                    Collection<StatusItem> items = null;
                    do {
                        items = getter.getAllStatusItems();
                        if (items == null) {
                            LOG.warning("Couldn't get StatusItems from StatusDB, sleeping...");
                            Thread.sleep(sleepTime *= 2);
                        }
                    } while (items == null);//wait until JMS on receiver end is started

                    for (StatusItem<?> item : items) {
                        update(item);
                    }
                } catch (Exception e) {
                    LOG.log(Level.SEVERE, e.getMessage(), e);
                }
                LOG.fine("End initItems");
            }
        });
    }

    private String substituteProperties(String url) {
        EProperties props = new EProperties();
        props.addAll(System.getProperties());
        props.put(CONF_DIR_PROPERTY, url);
        return props.get(CONF_DIR_PROPERTY, "/").toString();
    }

    public void stop() {

    }

    @Override
    public String getName() {
        return name;
    }

    /**
     * Create a StatusItem
     *
     * @param type    DataType of the item to create
     * @param newName name of the item
     * @param newVal  value of the item
     * @return a Some<StatusItem<?>> if everything is correct, otherwise a None.
     */
    private Optional<StatusItem<?>> createStatus(DataType type, String newName, String newVal) {
        StatusItem<?> newItem = null;
        try {
            switch (type) {
                case INT:
                    newItem = new BasicStatus<Integer>(newName, Integer.valueOf(newVal));
                    break;
                case FLOAT:
                    newItem = new BasicStatus<Float>(newName, Float.valueOf(newVal));
                    break;
                case DOUBLE:
                    newItem = new BasicStatus<Double>(newName, Double.valueOf(newVal));
                    break;
                case STRING:
                    newItem = new BasicStatus<String>(newName, newVal);
                    break;
                case HEALTH:
                    newItem = new HealthStatus(newName, Health.valueOf(newVal));
                    break;
            }
        } catch (NumberFormatException e) {
            LOG.log(Level.SEVERE, e.getMessage(), e);
        } catch (NullPointerException e) {
            LOG.log(Level.SEVERE, e.getMessage(), e);
        }
        if (newItem == null) {
            return Optional.absent();
        } else {
            return Optional.<StatusItem<?>>of(newItem);
        }
    }

    /**
     * Translate a StatusItem according to translations specified in the config file.
     *
     * @param item the item to translate
     * @param <T>
     * @return a Some<StatusItem<?>> with the translated item, or a None if a problem occured
     */
    protected <T> List<StatusItem<?>> translate(StatusItem<T> item) {
        List<StatusItem<?>> list = new ArrayList<StatusItem<?>>();
        for (String newName : names.get(item.getName())) {
            LOG.fine("Translating " + item);

            //translate
            String newVal = translations.get(Key.create(item.getName(),newName)).get(item.getValue().toString());

            //no translation, return default
            if (newVal == null) {
                if (defaults.get(Key.create(item.getName(),newName)) != null) {
                    Optional<StatusItem<?>> newItem = createStatus(types.get(Key.create(item.getName(),newName)), newName, defaults.get(Key.create(item.getName(),newName)));
                    if (newItem.isPresent()) {
                        list.add(newItem.get());
                    }
                }
            } else {
                //return proper translation
                Optional<StatusItem<?>> newItem = createStatus(types.get(Key.create(item.getName(),newName)), newName, newVal);
                if (newItem.isPresent()) {
                    list.add(newItem.get());
                }
            }
        }
        return list;
    }
}

File: gmp-status-translator/src/main/scala/edu/gemini/gmp/status/translator/AbstractStatusItemTranslator.scala

package edu.gemini.gmp.status.translator

import edu.gemini.aspen.giapi.status.Health
import edu.gemini.aspen.giapi.status.StatusItem
import edu.gemini.aspen.giapi.status.impl.BasicStatus
import edu.gemini.aspen.giapi.status.impl.HealthStatus
import edu.gemini.aspen.giapi.util.jms.status.StatusGetter
import edu.gemini.gmp.top.Top
import scala.concurrent.ExecutionContext.Implicits.global
import edu.gemini.gmp.status.translator.generated.{StatusType, DataType}
import java.io.{FileInputStream, File}
import edu.gemini.jms.api.{JmsProvider, JmsArtifact}

import scala.collection._
import scala.concurrent.Future

case class ItemTranslation[A, +B](destinationName: String, default: Option[B], translations: Map[A, B], destinationType: DataType) {
  def translate(value: A):StatusItem[_] = {
    val tr:TranslationType[_] = TranslationType.defaultTranslationType(destinationType)
    translateInternal(value)(tr)
  }
  private def translateInternal(value: A)(translation: TranslationType[_]):StatusItem[_] = {
    val d = default.getOrElse(translation.defaultValue)
    val defaultStatus = translation.toStatusItem(destinationName)(d.toString)
    translations.get(value).map(v => translation.toStatusItem(destinationName)(translation.toValue(v.toString).toString)).getOrElse(defaultStatus)
  }
}

trait TranslationType[A] {
  def toValue(value: String): A
  def toStatusItem(name: String)(value: String):StatusItem[A]
  def defaultValue: A
}

object TranslationType {
  implicit object IntTranslationType extends TranslationType[Int] {
    override def toValue(value: String):Int = value.toInt
    override def toStatusItem(name: String)(value: String) = new BasicStatus[Int](name, toValue(value))
    override val defaultValue = 0
  }
  implicit object DoubleTranslationType extends TranslationType[Double] {
    override def toValue(value: String):Double = value.toDouble
    override def toStatusItem(name: String)(value: String) = new BasicStatus[Double](name, toValue(value))
    override val defaultValue = 0.0
  }
  implicit object HealthTranslationType extends TranslationType[Health] {
    override def toValue(value: String):Health = Health.valueOf(value)
    override def toStatusItem(name: String)(value: String) = new HealthStatus(name, toValue(value))
    override val defaultValue = Health.BAD
  }
  implicit object StringTranslationType extends TranslationType[String] {
    override def toValue(value: String):String = value
    override def toStatusItem(name: String)(value: String) = new BasicStatus[String](name, toValue(value))
    override val defaultValue = ""
  }

  def toValue[A: TranslationType](value: String): A = implicitly[TranslationType[A]].toValue(value)

  def defaultTranslationType(statusType: DataType):TranslationType[_] = statusType match {
    case DataType.INT    => IntTranslationType
    case DataType.DOUBLE => DoubleTranslationType
    case DataType.HEALTH => HealthTranslationType
    case DataType.STRING => StringTranslationType
  }
}


abstract class AbstractStatusItemTranslator(top: Top, xmlFileName: String) extends StatusItemTranslator with JmsArtifact {

  val name = s"StatusItemTranslator: $this"
  var config: StatusItemTranslatorConfiguration = null
  val getter: StatusGetter = new StatusGetter("Status Translator initial item loader")

  val fileExistence = for {
    x <- Option(new File(xmlFileName))
    if x.exists()
  } yield new StatusItemTranslatorConfiguration(new FileInputStream(x))

  def translationMap(config: StatusType) = {
    import scala.collection.JavaConverters._
    import TranslationType._

    val fromTranslator = defaultTranslationType(config.getOriginalType)
    val toTranslator = defaultTranslationType(config.getTranslatedType)

    val result = for {
      item <- config.getMaps.getMap.asScala
    } yield toValue(item.getFrom)(fromTranslator) -> toValue(item.getTo)(toTranslator)
    result.toMap
  }

  def extract(config: StatusItemTranslatorConfiguration): List[(String, ItemTranslation[Any, Any])] =
    for {
      status      <- config.statuses
      ot           = status.getOriginalType
      dt           = status.getTranslatedType
      i            = TranslationType.defaultTranslationType(dt)
      default      = Option(status.getDefault).map(TranslationType.toValue(_)(i))
      translations = translationMap(status)
    } yield top.buildStatusItemName(status.getOriginalName) -> ItemTranslation(top.buildStatusItemName(status.getTranslatedName), default, translations, dt)

  val translations = fileExistence.map(extract).map(_.groupBy(a => a._1)).getOrElse(Map.empty)  //val translations = fileExistence.map(extract).map(_.toMap).getOrElse(Map.empty)

  override def startJms(provider: JmsProvider) {
    getter.startJms(provider)
    import scala.collection.JavaConversions._
    Future.apply {
      for {
        si <- getter.getAllStatusItems
      } update(si)
    }
  }

  override def stopJms {}

  def getName = name

  /**
   * Translate a StatusItem according to translations specified in the config file.
   *
   * @param item the item to translate
   * @return a Some<StatusItem<?>> with the translated item, or a None if a problem occured
   */
  def translate(item: StatusItem[_]): List[StatusItem[_]] = {
    val result = for {
        source <- translations.get(item.getName)
      } yield for {
          tr <- source
        } yield tr._2.translate(item.getValue)
    result.getOrElse(List.empty)
  }

}